StringRep.hs

```diff
@@ -13,6 +14,76 @@ import qualified Miso.String as MS
 import qualified Data.Text.Lazy.Builder as B
 import qualified Data.Text.Lazy as L
 import Terms
+import qualified Text.Earley as EP
+import qualified Text.Earley.Mixfix as EPM
+
+data Token = LParen | RParen | Word MS.MisoString | Dot | Binder MS.MisoString deriving (Show, Eq)
+data TermH = WordH (EPM.Holey Token) | BinderH Token | LamH TermH TermH | AppH TermH [TermH] deriving (Show, Eq)
+
+holey :: String -> EPM.Holey MS.MisoString
+holey ""       = []
+holey ('_':xs) = Nothing : holey xs
+holey xs       = Just (MS.toMisoString i) : holey rest
+  where (i, rest) = span (/= '_') xs
+
+mixfixTable :: [[(EPM.Holey MS.MisoString, EPM.Associativity)]]
+mixfixTable = (map . map) (first holey)
+  [ [("_->_",          EPM.RightAssoc)]
+  , [("_,_",           EPM.NonAssoc)]
+  , [("if_then_else_", EPM.RightAssoc)]
+  , [("_|-_:_",        EPM.NonAssoc)]
+  , [("_+_",           EPM.LeftAssoc)]
+  , [("_*_",           EPM.LeftAssoc)]
+  ]
+
+grammar :: EP.Grammar r (EP.Prod r Token Token TermH)
+grammar = mdo
+  identB    <- EP.rule $ BinderH <$> EP.satisfy isLegalBinder
+  identW    <- EP.rule $ (WordH . pure . Just) <$> EP.satisfy isLegalConst
+  atom      <- EP.rule $ identW
+                      <|> EP.namedToken LParen *> program <* EP.namedToken RParen
+  normalApp <- EP.rule $ atom
+                      <|> AppH <$> atom <*> some atom
+  mixfixApp <- EPM.mixfixExpression table normalApp (AppH . WordH)
+  program    <- EP.rule $ mixfixApp 
+                      <|> LamH <$> identB <*> program 
+  return program
+  where
+    table = map (map $ first $ map $ fmap (EP.namedToken . Word)) mixfixTable
+    illegalIdents = [s | xs <- mixfixTable , (ys, _) <- xs , Just s <- ys] -- TODO: switch to hashtable, but required hashable version for ghcjs and unordered-containers are conflicted. 
```

Comment:
This is not a big deal, it would only seriously hit performance if the list got very large